contents/javascript/Javascript-promises.md

```diff
@@ -90,57 +86,118 @@ These situations correspond to the 3 states of Promises in Javascript.
 
-function update(numTimes) {
-    return new Promise((resolve, reject) => {
-        var x = "I have not been updated";
-        setTimeout(function() {
-            x = "I have been updated " + numTimes + " times.";
-            console.log(x);
-            resolve(numTimes);
-        }, 1000);
-    });
-}
+The `async` keyword ensures that the function returns a promise, and as for `await`, you can think of it as waiting for some promise to return the value before continuing. So back to the same example.
 
-update(1).then((result) => update(result + 1))
-.then((result) => update(result + 1));
+So here is the code in the "promise style", as a recap.
 
+```javascript
+/** func implementation same as above **/
+// getData()
+// intenseWait(val)
+// filterData(data, callback)
+
+/** changes only required in main **/
+function main() {
+    return getData()
+        .then(data => filterData(data))
+        .then(fltered => console.log(filtered));
+}
+
+main();
 ```
 
-Now we update our function to take in a parameter `numTimes` so that we can log how many times `x` was updated. `then()` will return a Promise, which you can then again call `.then()`. By doing this, we can chain our Promises and ensure that execution of the code goes in the order that we require. In addition, it also allows your variables to be correctly passed down, ensuring that you will not have to deal with any undefined variables along the way.
+Using the "async await" style, only the `main()` function needs to be changed. After the change, you will realize that the code becomes much simpler to understand. There is a clear improvement in clarity and readability.
 
-Using Promises helps your code to be clean and easy to read, which makes it easy for anyone to instantly know what the code does, instead of having to trace through the code written with callbacks.
+```javascript
+async function main() {
+    const data = await getData();
+    const filtered = await filterData(data);
 
-## How Promises are used
+    console.log(filtered);
+}
+```
+
+`async` ensures that `main()` returns a promise. In our case, this will cause `main()` to return with `undefined` (equivalent to running `Promise.resolve()`). Another interesting thing to note: `await` will wait for the promise to return before executing anything below. In this case, `console.log` will be executed after the two `await` calls, even when it doesn't depend the results of those calls.
 
-There are many interesting ways to use Promises. Here we cover HTTP requests in Javascript, and Disk I/O in Node.js. No prior knowledge of Node.js is necessary.
+In the "promise" style, we handle errors using the `.catch()` block. However when using the "async await" style, we handle the errors using the more conventional `try ... catch` block. These can be explored further in the readings listed at the end of this post.
```

Comment:
Give a specific link right hear, instead of expecting the reader to visit many links to find the specific topic.